include/aws/mqtt/private/client_impl.h

```diff
@@ -85,77 +88,102 @@ struct aws_mqtt_client_connection {
 
     struct aws_mqtt_client *client;
 
-    /* The host information */
+    /* Channel handler information */
+    struct aws_channel_handler handler;
+    struct aws_channel_slot *slot;
+
+    /* The host information, changed by user when state is AWS_MQTT_CLIENT_STATE_DISCONNECTED */
     struct aws_string *host_name;
     uint16_t port;
     struct aws_tls_connection_options tls_options;
     struct aws_socket_options socket_options;
 
+    /* Connect parameters */
+    struct aws_byte_buf client_id;
+    bool clean_session;
+    uint16_t keep_alive_time_secs;
+    uint64_t request_timeout_ns;
+    struct aws_string *username;
+    struct aws_string *password;
+    struct {
+        struct aws_byte_buf topic;
+        enum aws_mqtt_qos qos;
+        bool retain;
+        struct aws_byte_buf payload;
+    } will;
+    struct {
+        uint64_t current;      /* seconds */
+        uint64_t min;          /* seconds */
+        uint64_t max;          /* seconds */
+        uint64_t next_attempt; /* milliseconds */
+    } reconnect_timeouts;
     /* User connection callbacks */
     aws_mqtt_client_on_connection_complete_fn *on_connection_complete;
     void *on_connection_complete_ud;
-    aws_mqtt_client_on_disconnect_fn *on_disconnect;
-    void *on_disconnect_ud;
     aws_mqtt_client_on_connection_interrupted_fn *on_interrupted;
     void *on_interrupted_ud;
     aws_mqtt_client_on_connection_resumed_fn *on_resumed;
     void *on_resumed_ud;
-
-    /* The state of the connection */
-    enum aws_mqtt_client_connection_state state;
-
-    /* Channel handler information */
-    struct aws_channel_handler handler;
-    struct aws_channel_slot *slot;
-
-    /* Keeps track of all open subscriptions */
-    struct aws_mqtt_topic_tree subscriptions;
     aws_mqtt_client_publish_received_fn *on_any_publish;
     void *on_any_publish_ud;
```

Comment:
This also feels like it should be synced, because delivery can be happening in one thread while this callback is set from another.

Reply:
hmmm, I thought about this. I cannot remember how I convinced myself we don't need a lock for this... Anyway, now I agree that we need put it into the synced_data, will do

Reply:
right! Here is the problem, user may probably call functions that grab a lock from callback, which will cause a deadlock... A workaround I can think of is do not allow user to change this on_publish callback when the state is CONNECTED.🤔

Reply:
That's a perfectly good solution.